### DISCUSSION TITLE
bytes in form is handled in a very strange way

### DISCUSSION THREAD

**Discussion by trim21** [encode/httpx#3168] [Potential Issue]
```python
import httpx


r = httpx.post(
    "https://httpbin.org/post",
    data={
        "content": bytes.fromhex(
            "4f5f13d16da9f10d013933673a1ef4e7b974dbe3"
        ),  # random non-printable bytes
    },
)
print(r.json()["form"])
```

and it didn't throw a exception, but just transform bytes with `str(b)`

related: https://github.com/encode/httpx/issues/1500 https://github.com/encode/httpx/pull/1539 https://github.com/encode/httpx/pull/1608

**Comment by lovelydinosaur** [comment]
That's not ideal behavior, is it?

That's occurring because we've got an all-purpose string coercion here.

https://github.com/encode/httpx/blob/4b85e6c3898b94e686b427afd83138c87520b479/httpx/_content.py#L136-L149

https://github.com/encode/httpx/blob/4b85e6c3898b94e686b427afd83138c87520b479/httpx/_utils.py#L56-L68

Presumably a better behavior here is to raise a `TypeError` for any form value that's not a string, and ensure the user is required to handle any coercion themselves?

There's actually a slightly wider design niggle here that I'd also like to pick up, but let's deal with one thing at a time.

**Reply by trim21** [reply]
I think it should raise a `TypeError` for bytes...

and for other types, maybe `enum.Enum` need speical treatment.

**Reply by lovelydinosaur** [reply]
If we were being strict, we'd *only* support string literals here, and nothing else.

Supportig ~JSON scalars, as we're currently doing is probably a reasonable compromise here...

* int
* float
* str
* bool (Represent as JSON-style "true"/"false")
* None (Represent as empty string)
* Enum (Use `.value` and then as above.) &mdash; *We're not handling this at the moment.*
* `TypeError` for anything else. &mdash; *We're not handling this at the moment, defaulting to coerce-to-str.*

Documenting what we support here and why is also *a thing*.

**Reply by trim21** [reply]
sounds reasonable to me

**Reply by lovelydinosaur** [reply]
Related https://github.com/encode/httpx/pull/2515

**Reply by trim21** [reply]
should we transfer this discuss to a issue first?

**Reply by lovelydinosaur** [reply]
I think so yes, though let's be sure we got it fully spec'ed...

* Are we confident with the resolution above?
* There's a behavioural change there wrt. not allowing silent str coercions, are we okay with that from perspective of existing users?
* Should we also be ensuring we've got same behaviour wrt. query params?

**Reply by trim21** [reply]
bump to 0.28.x, maybe?

**Reply by trim21** [reply]
I think your suggestion in https://github.com/encode/httpx/discussions/3168#discussioncomment-9128685 looks fine to me,

**Reply by trim21** [reply]
better to create a issue and discuss there. Too much replies in this thread.

**Reply by lovelydinosaur** [reply]
@trim21 - Okay would you like to open an issue for this?

We don't currently have a great place to write about this in the documentation, perhaps we can describe our API expectations to our users as part of the issue so that we've got something to later add into the docs.

Key points are...

* Form encoding is the same style that browsers use for uploading HTML forms.
* It either uses URL encoding or multipart encoding. 
* Only text is supported. This isn't a structured data type like JSON.
* For convenience we support handling the following primitives [...]. Any other data needs to be coerced to a str.

**Reply by trim21** [reply]
https://github.com/encode/httpx/issues/3170
